libnmstate/validator.py

```diff
@@ -48,6 +48,10 @@ class NmstateRouteWithNoIPInterfaceError(NmstateValueError):
     pass
 
 
+class NmstateDuplicateInterfaceNameError(NmstateValueError):
```

Comment:
IMHO this is a very specific error that does not need its own exception. The explanation could be added to the exception message.

Reply:
Also public exceptions belong in errors.py.

Reply:
The error name explains what it is, which is better than text.
It inherits from a public error: `NmstateValueError`. Therefore, the caller will catch it if it specify it.

This pattern is already in use and is very helpful to detect the source of the problem without overloading the public error list. If someone will ask, we can promote it to be public easily.

Reply:
If something catches the value error and then presents it to the user, it might look like:

`Bad values in desired state: foo0, foo0`

or the user needs to specifically catch for `NmstateDuplicateInterfaceNameError` to change this into a message for the user and for this it becomes part of the API.

Reply:
If you are trying to claim that an exception text is better than a exception type, we are in a disagreement here.
The caller can print the error name + text easily.

If we are interested in exposing these exception classes publicly, it is an easy step and it will not touch the production code. If we need to go the other way around it is harder.

Reply:
The internal class name does not belong in an error message meant for a user, this is something for developers. Also it would not be obvious for a developer using libnmstate who adds code to except a ValueError that they need to print the class name because it contains the actual information for the user instead of the exception message.

Except for the route errors that you added, all other places in Nmstate specify the problem for a value error in the message, for example:
https://github.com/nmstate/nmstate/blob/master/libnmstate/nm/ovs.py#L69

This should happen in a uniform way. And IMHO the the exception type is meant for the code to allow to run a different code path in case of an exception (for example a FileNotFoundError might trigger a different code path than a PermissionDeniedError) it does not seem to be a good idea to introduce new exceptions.

Reply:
> The internal class name does not belong in an error message meant for a user, this is something for developers. Also it would not be obvious for a developer using libnmstate who adds code to except a ValueError that they need to print the class name because it contains the actual information for the user instead of the exception message.

This is highly opinionated.
Current change uses an existing pattern.
If you want to challenge this pattern, please suggest it though its own PR.

> This should happen in a uniform way. And IMHO the the exception type is meant for the code to allow to run a different code path in case of an exception (for example a FileNotFoundError might trigger a different code path than a PermissionDeniedError) it does not seem to be a good idea to introduce new exceptions.

Again, this is opinionated. There is nothing wrong to raise exceptions that inherit from an API one. The applications that use the lib may do it as well.
If the client of the lib wants to re-use the error text, they need to parse it anyway. Taking the info from the text-only, from the class+text (by serializing the exception using `str` or `logging.except`) is up to the user.

I prefer being specific with a type while you suggest using a more generic type and add text.
In the unit test I could use this exception for the tests and still do not have to expose it through the formal API.

Will adding descriptive text to the specific exception be a good compromise for you?